README.md
# EasyTextInput

An easy to use, complete text input brought to you by Brainn.co

### Installation

Add this to your pubspec.yaml (or create it):

```yaml
dependencies:
easy_text_input: ^1.0.2
```

Then run the flutter tooling:

```bash
flutter packages get
```

## Demo

![demo](https://github.com/CarlosMion/Gifs-and-Images/blob/master/easyTextInput.gif)

## How to use

You will need a Bloc with a stream to controll the data flowing through the InputField, for a better explanation on how blocs work please check the example folder.

#### Simple Field

```dart
InputField(
              stream: exampleBloc.exampleStream,
              labelText: 'Example',
              onChanged: exampleBloc.sinkExample,
            ),
```

#### Example obscured Field

```dart
InputField(
              stream: exampleBloc.password,
              labelText: 'Password',
              onChanged: exampleBloc.sinkPassword,
              obscureText: true,
              successIcon: _successIcon,
              errorIcon: _errorIcon,
              obscureTextIconOff: EyeIcon.close,
              obscureTextIconOn: EyeIcon.open,
            ),
```

#### Example masked Field

```dart
InputField(
              stream: exampleBloc.phone,
              labelText: 'Phone',
              obscureText: true,
              onChanged: exampleBloc.sinkPhone,
              controller: MaskedTextController(mask: '(00) 00000-00000'),
              hintText: '(31) 91234-5678',
              successIcon: _successIcon,
            ),
```

## Attributes

Customizable attributes for InputField

<table>
    <th>Attribute Name</th>
    <th>Type</th>
    <th>Required</th>
    <th>Default Value</th>
    <th>Example Value</th>
    <th>Description</th>
    <tr>
        <td>stream</td>
        <td>Stream</td>
        <td>true</td>
        <td></td>
        <td>exampleBloc.emailStream</td>
        <td>The stream that controls the data flowing through the InputField Widget</td>
    </tr>
    <tr>
        <td>onChanged</td>
        <td>Function</td>
        <td>true</td>
        <td></td>
        <td>exampleBloc.sinkEmail</td>
        <td>The method that will deal with every change in the value of the focused InputField</td>
    </tr>
    <tr>
        <td>labelText</td>
        <td>String</td>
        <td>false</td>
        <td>''</td>
        <td>'E-mail'</td>
        <td>The Label of the InputField, The information that will float to the top of the InputField when the Widget is focused</td>
    </tr>
    <tr>
        <td>hintText</td>
        <td>String</td>
        <td>false</td>
        <td>''</td>
        <td>'hint@example.com'</td>
        <td>The text that will show only when the InputField is focused, to show an example of how the input should be</td>
    </tr>
    <tr>
        <td>counterText</td>
        <td>String</td>
        <td>false</td>
        <td>''</td>
        <td>'1/6'</td>
        <td>a small text that shows on the inferior right of the field to show the user a additional information, like a number to show the user how many characters he has typed</td>
    </tr>
    <tr>
        <td>obscureText</td>
        <td>bool</td>
        <td>false</td>
        <td></td>
        <td>true</td>
        <td>Controls whether the text should show or be obscured, mainly used for password fields</td>
    </tr>
    <tr>
        <td>controller</td>
        <td>dynamic</td>
        <td>false</td>
        <td></td>
        <td>MaskedTextController(mask: '(00) 00000-00000')</td>
        <td>The prop that will connect to the Flutter's native TextInput controller prop, a commom use for this particular prop is for controlling masks on the Text</td>
    </tr>
    <tr>
        <td>keyboardType</td>
        <td>TextInputType</td>
        <td>false</td>
        <td>TextInputType.text</td>
        <td>TextInputType.number</td>
        <td>Controls the type of the keyboard that will show to the user when the InputField is focused</td>
    </tr>
    <tr>
        <td>maxLength</td>
        <td>int</td>
        <td>false</td>
        <td></td>
        <td>10</td>
        <td>Set the maximum amount of characters the InputField will accept</td>
    </tr>
    <tr>
        <td>errorIcon</td>
        <td>Widget</td>
        <td>false</td>
        <td>const SizedBox.shrink()</td>
        <td>ImageIcon();</td>
        <td>The icon that will show at the end of the InputField showing that the input is invalid</td>
    </tr>
    <tr>
        <td>successIcon</td>
        <td>Widget</td>
        <td>false</td>
        <td>const SizedBox.shrink()</td>
        <td>ImageIcon();</td>
        <td>The icon that will show at the end of the InputField showing that the input is valid</td>
    </tr>
     <tr>
        <td>obscureTextIconOff</td>
        <td>Widget</td>
        <td>false</td>
        <td>const SizedBox.shrink()</td>
        <td>ImageIcon();</td>
        <td>The icon that will show at the end of the InputField showing that the obscureText feature is off</td>
    </tr>
    <tr>
        <td>obscureTextIconOn</td>
        <td>Widget</td>
        <td>false</td>
        <td>const SizedBox.shrink()</td>
        <td>ImageIcon();</td>
        <td>The icon that will show at the end of the InputField showing that the obscureText feature is on</td>
    </tr>
    <tr>
        <td>borderRadius</td>
        <td>double</td>
        <td>false</td>
        <td>8.0</td>
        <td>12.0</td>
        <td>Controls the curvature of the borders of the InputField</td>
    </tr>
    <tr>
        <td>successColor</td>
        <td>Color</td>
        <td>false</td>
        <td>Colors.green</td>
        <td>Colors.blue</td>
        <td>Sets the color that the InputField will be colored with when the input is considered valid</td>
    </tr>
    <tr>
        <td>errorColor</td>
        <td>Color</td>
        <td>false</td>
        <td>Colors.red</td>
        <td>Colors.purple</td>
        <td>Sets the color that the InputField will be colored with when the input is considered invalid</td>
    </tr>
</table>


CHANGELOG.md
## [1.0.0] - 23/03/2020

* Launched the Complete, fully customisable, and non-verbose InputField for flutter apps.

## [1.0.2] - 05/05/2020

* updated library homepage


test/input_field_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:easy_text_input/easy_text_input.dart';

Widget _buildTestableWidget({Widget child}) {
  return MaterialApp(
    home: Scaffold(
      body: child,
    ),
  );
}

void main() {
  testWidgets('shows input field', (WidgetTester tester) async {
    await tester.pumpWidget(
      _buildTestableWidget(
        child: InputField(
          stream: null,
          onChanged: null,
          labelText: 'anything',
        ),
      ),
    );

    await tester.pump();

    expect(find.byType(InputField), findsOneWidget);
    expect(find.text('anything'), findsOneWidget);
  });
}


lib/input_field.dart
library easy_text_input;

import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

class InputField extends StatefulWidget {
  const InputField({
    @required this.stream,
    @required this.onChanged,
    this.labelText = '',
    this.hintText = '',
    this.counterText = '',
    this.obscureText,
    this.controller,
    this.keyboardType = TextInputType.text,
    this.maxLength,
    this.errorIcon = const SizedBox.shrink(),
    this.successIcon = const SizedBox.shrink(),
    this.obscureTextIconOff = const SizedBox.shrink(),
    this.obscureTextIconOn = const SizedBox.shrink(),
    this.borderRadius = 8,
    this.successColor = Colors.green,
    this.errorColor = Colors.red,
  });

  final Stream stream;
  final Function onChanged;
  final String labelText;
  final String hintText;
  final String counterText;
  final bool obscureText;
  final dynamic controller;
  final TextInputType keyboardType;
  final int maxLength;
  final Widget errorIcon;
  final Widget successIcon;
  final Widget obscureTextIconOn;
  final Widget obscureTextIconOff;
  final double borderRadius;
  final Color successColor;
  final Color errorColor;

  @override
  State<StatefulWidget> createState() {
    return InputFieldState(
      obscureText: obscureText,
    );
  }
}

class InputFieldState extends State<InputField> {
  InputFieldState({@required this.obscureText}) {
    inputFieldFocusNode.addListener(_onFocus);
  }

  bool obscureText;
  bool showFieldBorder = false;
  final FocusNode inputFieldFocusNode = FocusNode();

  static bool _isValidSnapshot(
    AsyncSnapshot<String> snapshot, [
    bool showBorder = false,
  ]) {
    if (showBorder) {
      if (snapshot.hasData && !snapshot.hasError) {
        return true;
      }
      if (snapshot.hasError) {
        return false;
      }
    }
    return null;
  }

  void _alternateObscureText() {
    setState(() {
      obscureText = !obscureText;
    });
  }

  void _onFocus() {
    final bool hasFocus = inputFieldFocusNode.hasFocus;
    Scaffold.of(context).removeCurrentSnackBar();
    setState(() {
      showFieldBorder = !hasFocus;
    });
  }

  @override
  Widget build(BuildContext context) {
    return StreamBuilder<String>(
        stream: widget.stream,
        builder: (
          BuildContext context,
          AsyncSnapshot<String> snapshot,
        ) {
          return Container(
            child: TextField(
              keyboardType: widget.keyboardType,
              maxLength: widget.maxLength,
              focusNode: inputFieldFocusNode,
              controller: widget.controller,
              obscureText: obscureText ?? false,
              onChanged: widget.onChanged,
              decoration: InputDecoration(
                counterText: widget.counterText,
                labelStyle: _showLabelStyle(snapshot),
                labelText: widget.labelText,
                hintText: widget.hintText,
                errorText: showFieldBorder ? snapshot.error : null,
                suffixIcon: _showIcon(snapshot),
                enabledBorder: _showBorder(snapshot),
                border: OutlineInputBorder(
                  borderRadius: BorderRadius.all(
                    Radius.circular(widget.borderRadius),
                  ),
                ),
              ),
            ),
          );
        });
  }

  Widget _showIcon(AsyncSnapshot<String> snapshot) {
    if (_isValidSnapshot(snapshot, showFieldBorder) == null) {
      if (obscureText == null) {
        return SizedBox.shrink();
      }

      return GestureDetector(
        child:
            obscureText ? widget.obscureTextIconOff : widget.obscureTextIconOn,
        onTap: _alternateObscureText,
      );
    }

    return _isValidSnapshot(snapshot, showFieldBorder)
        ? widget.successIcon
        : widget.errorIcon;
  }

  TextStyle _showLabelStyle(AsyncSnapshot<String> snapshot) {
    if (_isValidSnapshot(snapshot, showFieldBorder) == null) {
      return null;
    }

    return _isValidSnapshot(snapshot, showFieldBorder)
        ? TextStyle(
            color: widget.successColor, //green
          )
        : null;
  }

  OutlineInputBorder _showBorder(AsyncSnapshot<String> snapshot) {
    if (_isValidSnapshot(snapshot, showFieldBorder) == null) {
      return null;
    }
    return _isValidSnapshot(snapshot, showFieldBorder)
        ? OutlineInputBorder(
            borderSide: BorderSide(
              color: widget.successColor, //green
              width: 1.0,
            ),
          )
        : OutlineInputBorder(
            borderSide: BorderSide(
              color: widget.errorColor, //red
              width: 1.0,
            ),
          );
  }
}


example/lib/blocs/validators.dart
import 'dart:async';

class Validators {
  static StreamTransformer<String, String> validateInputHasMin(int size,
      [String message]) {
    return StreamTransformer<String, String>.fromHandlers(handleData: (
      String value,
      EventSink<String> sink,
    ) {
      if (value == null || value.isEmpty) {
        sink.add(null);
        return;
      }
      if (value.length >= size) {
        sink.add(value);
      } else {
        sink.addError(message);
      }
    });
  }

  static StreamTransformer<String, String>
      validateInputHasExactNumberOfCharacters(int characters,
          [String message]) {
    return StreamTransformer<String, String>.fromHandlers(handleData: (
      String value,
      EventSink<String> sink,
    ) {
      if (value == null || value.isEmpty) {
        sink.add(null);
        return;
      }

      if (value.length == characters) {
        sink.add(value);
      } else {
        sink.addError(message);
      }
    });
  }
}


example/lib/blocs/example_bloc.dart
import 'package:rxdart/rxdart.dart';

import 'validators.dart';

class ExampleBloc {
  final phoneCharactersAmount = 15;
  final minPasswordCharacters = 5;

  final BehaviorSubject<String> _phone = BehaviorSubject<String>();
  final BehaviorSubject<String> _password = BehaviorSubject<String>();

  Stream<String> get phone => _phone.stream.transform(
      Validators.validateInputHasExactNumberOfCharacters(phoneCharactersAmount,
          'The phone number must have exaclty 11 numbers'));

  Stream<String> get password =>
      _password.stream.transform(Validators.validateInputHasMin(
          minPasswordCharacters,
          'The password must have at least 5 characters'));

  Function get sinkPhone => _phone.sink.add;

  Function get sinkPassword => _password.sink.add;

  String get phoneValue => _phone.value;

  String get passwordValue => _password.value;

  void dispose() {
    _phone.close();
    _password.close();
  }
}


example/lib/screens/example_screen.dart
import 'package:easy_text_input/easy_text_input.dart';
import 'package:example/widgets/eye_icon.dart';
import 'package:flutter/material.dart';
import 'package:flutter_masked_text/flutter_masked_text.dart';

import '../blocs/example_bloc.dart';

class ExampleScreen extends StatefulWidget {
  const ExampleScreen({
    @required this.exampleBloc,
  });

  final ExampleBloc exampleBloc;

  @override
  State<StatefulWidget> createState() {
    return ExampleScreenState(
      exampleBloc: exampleBloc,
    );
  }
}

class ExampleScreenState extends State<ExampleScreen> {
  ExampleScreenState({
    @required this.exampleBloc,
  });

  final ExampleBloc exampleBloc;

  final ImageIcon _errorIcon = ImageIcon(
    AssetImage('lib/assets/error_circle.png'),
    color: Colors.red,
  );

  final ImageIcon _successIcon = ImageIcon(
    AssetImage('lib/assets/check_circle.png'),
    color: Colors.green,
  );

  @override
  Widget build(BuildContext context) {
    return Container(
        margin: const EdgeInsets.all(16),
        padding: EdgeInsets.only(top: 120),
        child: Column(
          children: <Widget>[
            InputField(
              stream: exampleBloc.password,
              labelText: 'Password',
              onChanged: exampleBloc.sinkPassword,
              obscureText: true,
              successIcon: _successIcon,
              errorIcon: _errorIcon,
              obscureTextIconOff: EyeIcon.close,
              obscureTextIconOn: EyeIcon.open,
            ),
            SizedBox(
              height: 16,
            ),
            InputField(
              stream: exampleBloc.phone,
              labelText: 'Phone',
              obscureText: false,
              onChanged: exampleBloc.sinkPhone,
              controller: MaskedTextController(mask: '(00) 00000-00000'),
              hintText: '(31) 91234-5678',
              successIcon: _successIcon,
            ),
          ],
        ));
  }
}
